qmi/core/transport.py

```diff
@@ -858,70 +853,105 @@ def read(self, nbytes: int, timeout: Optional[float]) -> bytes:
+                ret = self._read_buffer
+                self._read_buffer = bytes()
+                return ret
 
-        data = self._read_buffer[:nbytes]
-        self._read_buffer = self._read_buffer[nbytes:]
-        return data
+            # Read buffer was of wrong length or is empty - read a new message from the instrument.
+            self._read_buffer += self._read_message(timeout)
+            nbuf = len(self._read_buffer)
+            if nbuf != nbytes:
+                _logger.debug("USBTMC read buffer contained data %s" % self._read_buffer.decode())
+                self._read_buffer = bytes()
+                raise QMI_EndOfInputException(
+                    f"The read buffer did not contain expected bytes of data ({nbuf} != {nbytes}."
+                )
 
     def read_until(self, message_terminator: bytes, timeout: Optional[float]) -> bytes:
-        """Read a single USBTMC message from the instrument."
+        """The specified message_terminator is ignored. Instead, the instrument must autonomously indicate the end
+        of the message according to the USBTMC protocol.
 
-        The specified message_terminator is ignored.
-        Instead, the instrument must autonomously indicate the end of the
-        message according to the USBTMC protocol.
+        As the message_terminator is not used, we forward simply the call to the `read_until_timeout` call.
 
-        This method blocks until a complete USBTMC message is available.
+        Parameters:
+            message_terminator: This input is ignored.
+            timeout: Maximum time to wait (in seconds).
+
+        Returns:
+            Received bytes.
+        """
+        return self.read_until_timeout(0, timeout)
+
+    def read_until_timeout(self, nbytes: int, timeout: float) -> bytes:
+        """Read a single USBTMC message from the instrument.
+
+        If there is already data in the buffer, return it.
 
         If the timeout expires before the message is received, the read is
         aborted and any data already received are discarded. In this
         case QMI_TimeoutException is raised.
 
         Parameters:
-            message_terminator: Message terminator characters.
+            nbytes: This input is ignored.
             timeout: Maximum time to wait (in seconds).
 
         Returns:
-            Received bytes, including the terminator.
+            Received bytes.
 
         Raises:
             ~qmi.core.exceptions.QMI_TimeoutException: If the timeout expires before the
-                requested number of bytes are available.
-
+            requested number of bytes are available.
         """
         self._check_is_open()
 
         # USB requires a timeout
         if timeout is None:
             timeout = self.DEFAULT_READ_TIMEOUT
 
+        data = bytes()
         if self._read_buffer:
             # Data already available in buffer; return it now.
-            data = self._read_buffer
+            data += self._read_buffer
```

Comment:
really minor: but what was reason for adding the `data` and `+=`?

Reply:
Yes, you're right, it is not needed as we take over a (byte) string and not something mutable like a list. So doing `self._read_buffer = bytes()` does not clear the overtaken string (`data = self._read_buffer`) by reference. I will change it back.

Reply:
In the end I would suggest to remove this, w.r.t. comment above about removing `self._read_buffer` for the USBTMC transport altogether.